### PR TITLE
feat(resharding) - load memtries pre-resharding

### PR DESCRIPTION
**Pull request by wacban**
The shard trie must be loaded into memory before it can be split in resharding. As of today some nodes do not enable memtrie so it is necessary to load the trie into memory on demand before resharding. This is the implementation of option 2, loading the shards into memory on startup, for all shards that will be resharding in this client / binary.

### CODE REVIEW

**Comment by wacban** [comment]
I added testloop test for now and checked that it fails without this feature and passes with it. I still owe some proper unit tests, I will add those tomorrow.

File: chain/epoch-manager/src/adapter.rs [hunk 1]
```diff
@@ -644,6 +649,14 @@ impl EpochManagerAdapter for EpochManagerHandle {
         self.get_shard_layout(&epoch_id)
     }
 
+    fn get_shard_layout_from_protocol_version(
+        &self,
+        protocol_version: ProtocolVersion,
+    ) -> Result<ShardLayout, EpochError> {
+        let epoch_manager = self.read();
+        Ok(epoch_manager.get_epoch_config(protocol_version).shard_layout)
```
Comment (Trisfald) [inline chain/epoch-manager/src/adapter.rs:657]: Can we change `get_shard_layout_from_protocol_version` to return a `ShardLayout`?

The version in `kv-runtime` shouldn always succeed

File: chain/chain/src/resharding/manager.rs [hunk 2]
```diff
@@ -103,6 +105,63 @@ impl ReshardingManager {
...
+                    break;
+                };
+                if children.len() > 1 {
+                    result.insert(shard_uid);
+                    break;
+                }
+            }
+        }
+
+        // Ok([].into())
```
Comment (Trisfald) [inline chain/chain/src/resharding/manager.rs:161]: nit: leftover

**Review by Trisfald — COMMENTED** [review]
Looking good so far, waiting for the tests of `get_extra_load_mem_tries_for_shards` 👍

File: chain/chain/src/resharding/manager.rs [hunk 3]
```diff
@@ -103,6 +105,63 @@ impl ReshardingManager {
...
+        let head_shard_layout =
+            self.epoch_manager.get_shard_layout_from_protocol_version(head_protocol_version)?;
+        for shard_uid in head_shard_layout.shard_uids() {
+            let shard_id = shard_uid.shard_id();
+            for shard_layout in &shard_layouts {
+                let children = shard_layout.get_children_shards_uids(shard_id);
+                let Some(children) = children else {
+                    break;
+                };
+                if children.len() > 1 {
```
Comment (staffik) [inline chain/chain/src/resharding/manager.rs:154]: sanity check, it can only be 0 or 2, right?

File: core/store/src/trie/shard_tries.rs [hunk 4]
```diff
@@ -445,21 +446,35 @@ impl ShardTries {
...
         parallelize: bool,
     ) -> Result<(), StorageError> {
         let trie_config = &self.0.trie_config;
         let shard_uids_to_load = tracked_shards
             .iter()
-            .copied()
             .filter(|shard_uid| {
                 trie_config.load_mem_tries_for_tracked_shards
                     || trie_config.load_mem_tries_for_shards.contains(shard_uid)
+                    || extra_load_mem_tries_for_shards.contains(shard_uid)
```
Comment (staffik) [inline core/store/src/trie/shard_tries.rs:466]: I have a concern about it:
If the resharding is planned 2 epochs later, and we do not care about parent shard in this or next epoch, then the parent shard won't be contained in `tracked_shards` and won't be loaded?

Comment (staffik) [reply core/store/src/trie/shard_tries.rs:466]: I mean, `tracked_shards` is already filtered with:
```
let tracked_shards: Vec<_> = shard_uids
            .iter()
            .filter(|shard_uid| {
                shard_tracker.care_about_shard(
```

Comment (shreyan-gupta) [reply core/store/src/trie/shard_tries.rs:466]: Hmm, agree with this. While this solution works for the case we are trying to handle, i.e. RPC nodes that track all shards, this isn't a general solution.

What if instead we call some async version of this function (like the state sync one) at the epoch boundary (first block of epoch)? @staffik, would that work?

Comment (staffik) [reply core/store/src/trie/shard_tries.rs:466]: I worry loading memtrie for a currently active shard might not work out of the box.
In this case, I think something in the lines of `shard_uids.filter(extra_load_mem_tries_for_shards` would just work.

Sanity check: if we gonna track child shard but we do not track parent shard, we state sync to parent shard and do we load memtrie after state sync?

Comment (Trisfald) [reply core/store/src/trie/shard_tries.rs:466]: From another point of view, we could assert that the node is configured to track all shard if `extra_load_mem_tries_for_shards` is used and/or non empty.

It is nice to have a simple logic even if the solution applies only to RPC nodes (it's tailored for them after all)

Comment (wacban) [reply core/store/src/trie/shard_tries.rs:466]: Yeah good point but I think it should be covered by the state sync integration. 

case a
Let's say a validator node doesn't track shard S on startup. A few epochs later it is assigned shard S. I think at this point it should do the same check as here, whether S will be resharded within this client and if so load it into memory. 

case b
Let's say a validator node doesn't track shard S on startup. A few epochs later it is assign a child of shard S, called S'. Let's say it's the first epoch where S' exists. Here the node should get S in the current epoch, load it into memory and perform resharding to get S'. 

cc @marcelo-gonzalez thoughts? 

@staffik and @shreyan-gupta Are you happy to keep this PR as is and cover the other cases separately? I think some code may be reusable but we can just refactor it as part of the state sync work.

File: chain/chain/src/resharding/manager.rs (hunk 3 above)
Comment (wacban) [reply chain/chain/src/resharding/manager.rs:154]: Today it can only be 1 or 2. 1 means no spit, the shard stays as is and 2 mean split into two children. In the future we'll have plenty of fun with merging though ;)

File: core/store/src/trie/shard_tries.rs (hunk 4 above)
Comment (shreyan-gupta) [reply core/store/src/trie/shard_tries.rs:466]: Aah wait, one thing I don't understand, currently we don't enable memtries for RPC nodes right? Where is that handled in the codebase? I don't see any "if rpc" conditions in the `load_mem_tries_for_enabled_shards` or any other function.

Also, like Wac mentioned, for normal nodes that didn't track S or S', state sync should be able to handle it without issues and we don't need to do anything extra. This PR should probably only address RPC nodes (that track all shards)?

Comment (staffik) [reply core/store/src/trie/shard_tries.rs:466]: > we don't enable memtries for RPC nodes right? Where is that handled in the codebase?

It is handled in `config.json` by RPC operator, by setting `load_mem_tries_for_tracked_shards: false`

Comment (wacban) [reply core/store/src/trie/shard_tries.rs:466]: > Aah wait, one thing I don't understand, currently we don't enable memtries for RPC nodes right? Where is that handled in the codebase? I don't see any "if rpc" conditions in the load_mem_tries_for_enabled_shards or any other function.

I believe RPC can decide for itself what shards it wants to load by setting the appropriate configuration. 

I recall validators should be forced to enable `load for tracked` config, there should be an assertion somewhere, @staffik will know where it is.

Comment (staffik) [reply core/store/src/trie/shard_tries.rs:466]: @wacban Did you mean this: https://github.com/near/nearcore/blob/master/chain/client/src/client_actor.rs#L330-L334 ?

Comment (wacban) [reply core/store/src/trie/shard_tries.rs:466]: heh, I did mean that line but it doesn't do what I expected. This one is for single shard tracking. Do we not force validators to use memtrie today?

Comment (staffik) [reply core/store/src/trie/shard_tries.rs:466]: I think we do not force them programatically, though we tell them to enable memtries, otherwise they are likely to fall behind after we changed gas cost or sth like that

**Comment by shreyan-gupta** [comment]
I'm leaving final review to @staffik! :)

**Review by staffik — APPROVED** [review]
> keep this PR as is and cover the other cases separately

I am happy to keep this PR as is and cover the other cases separately (if there is anything to cover).

**Comment by wacban** [comment]
> Looking good so far, waiting for the tests of `get_extra_load_mem_tries_for_shards` 👍

Done, after moving `get_shard_uids_pending_resharding` to epoch manager.

File: integration-tests/src/test_loop/tests/resharding_v3.rs
Comment (Trisfald) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:1]: Are these changes needed in this PR? :)

File: core/store/src/trie/shard_tries.rs (hunk 4 above)
Comment (marcelo-gonzalez) [reply core/store/src/trie/shard_tries.rs]: > cc @marcelo-gonzalez thoughts?
> 

Yeah I think that description is right. Case b is actually the motivation behind this PR: https://github.com/near/nearcore/pull/12404
